workflows/general/fastq-to-ubam.wdl

```diff
@@ -0,0 +1,116 @@
+            help: "See ../../data_structures/read_group.wdl for additional documentation.",
+            external_help: "https://samtools.github.io/hts-specs/SAMv1.pdf",
+        }
+        prefix: "Prefix for the BAM file. The extension `.bam` will be added."
+        validate_input: "Ensure input BAM is well-formed before beginning harmonization?"
+        use_all_cores: "Use all cores? Recommended for cloud environments."
+    }
+
+    input {
+        Array[File] read_one_fastqs_gz
+        Array[File] read_two_fastqs_gz
+        Array[ReadGroup] read_groups
+        String prefix = basename(read_one_fastqs_gz[0], ".fastq.gz")
+        Boolean validate_input = true
+        Boolean use_all_cores = false
+    }
+
+    call parse_input { input:
+        read_one_fastqs_gz,
+        read_two_fastqs_gz,
+        read_groups,
+    }
+
+    scatter (read_group in read_groups) {
+        call read_group.validate_read_group as validate_readgroups { input:
+            read_group,
+            required_fields = ["ID", "SM", "LB", "CN", "DT"]
+        }
+    }
+
+    call ubam_core.fastq_to_ubam_core after parse_input after validate_readgroups { input:
+        read_one_fastqs_gz,
+        read_two_fastqs_gz,
+        read_groups,
+        prefix,
+        validate_input,
+        use_all_cores,
+    }
+
+    output {
+        File unaligned_bam = fastq_to_ubam_core.unaligned_bam
+    }
+}
+
+task parse_input {
+    meta {
+        description: "Parses and validates the `fastq-to-ubam` workflow's provided inputs"
+        outputs: {
+            check: "Dummy output to indicate success and to enable call-caching"
+        }
+    }
+
+    parameter_meta {
+        read_one_fastqs_gz: "Array of gzipped FASTQ files with 1st reads in pair"
+        read_two_fastqs_gz: "Array of gzipped FASTQ files with 2nd reads in pair"
+        read_groups: {
+            description: "An Array of structs defining read groups to include in the harmonized BAM. Must correspond to input FASTQs. Each read group ID must be contained in the basename of a FASTQ file or pair of FASTQ files if Paired-End. This requirement means the length of `read_groups` must equal the length of `read_one_fastqs_gz` and the length of `read_two_fastqs_gz` if non-zero. Only the `ID` field is required, and it must be unique for each read group defined. See top of file for help formatting your input JSON.",  # Does not currently handle unknown RG case
+            help: "See ../../data_structures/read_group.wdl for additional documentation.",
+            external_help: "https://samtools.github.io/hts-specs/SAMv1.pdf",
+        }
+    }
+
+    input {
+        Array[File] read_one_fastqs_gz
+        Array[File] read_two_fastqs_gz
+        Array[ReadGroup] read_groups
+    }
+
+    command <<<
```

Comment:
This looks like it could be "outsourced" to `sort_star_input.py` for some more robust checks? Just an idea

Reply:
Wouldn't that require passing the full fastq files to this step? That or we'd need to add a step somewhere in the workflow to run that script. We'd also probably need to rename that script if we did that.

Reply:
`sort_star_input.py` just works on the paths, not the actual files. I think `Array[File]` _should_ coerce to `Array[String]` "as expected", but we'd need some testing

Reply:
OK, I added that. I didn't remove the old checks, just supplemented them. It's a little odd now that the `parse_input` task in `fastq-to-ubam` uses the `star` Docker image, but I'm not sure that's a big deal. We'll just have to be careful when bumping image versions.

Reply:
I guess it's fine for now, but can you open an issue for refactoring this? I think the best set up would probably be rename and move `sort_star_input.py` into the `util` image, and then do the STAR input checking as part of `parse_input` instead of as part of the STAR task? I think that would also remove the need to host our own STAR image, wouldn't it?

Reply:
Yes, that sounds like a good improvement: https://github.com/stjudecloud/workflows/issues/222